### PR TITLE
[spi_device] Block the passthrough with BUSY status

### PR DESCRIPTION
**Pull request by eunchan**
This commit addresses the issue #10785.

The BUSY @ SYS_CLK is set when SPI_DEVICE receives commands that satisfy:

- busy field in the CMD_INFO of the opcode is set
- CSb is de-asserted (SPI transaction completion)

When BUSY is set, SW may process the received command (assuming upload
field is set). After the command is processed by SW, SW may clear the
BUSY @SYS_CLK, which then will be propagated into SPI CLK (SCK) domain
at the next SPI transaction.

This commit is to turn off the passthrough interface if BUSY is set. So,
that the SW does not have to manually turn off the passthrough.

Keep in mind that this change cannot fully cover the corner cases:

- CSb is about to asserted (Beginning of the SPI transaction) at the
  same time BUSY is cleared by SW.

It is due to the CDC path of CSb to be used in SYS_CLK domain. It needs
to pass through 2FF synchronizer to be used in SYS_CLK safely. The delay
is 2 SYS_CLK. If BUSY falls in that window, the BUSY clearing may
unblock passthrough while SPI is active. The behavior of the IP is
undetermined.

### CODE REVIEW

File: hw/ip/spi_device/data/spi_device.hjson [hunk 1]
```diff
@@ -517,7 +517,8 @@
       fields: [
         { bits: "0"
           name: "busy"
-          desc: "BUSY"
+          desc: '''BUSY signal is cleared when CSb is high. SW should read
+            back the register to confirm the value is cleared.'''
```
Comment (a-will) [inline hw/ip/spi_device/data/spi_device.hjson:521]: Did you have a situation in mind that motivates software's need for confirmation?

While the BUSY bit is set in the CSR and/or the BUSY bit in the SPI status register has a clearing event pending, passthrough should be disabled. However, either the host should not use this register's value as an indicator that `spi_host` can be used, or the CSR bit should update in the sys_clk domain immediately, right? Does software care when the clearing event "transfers" to the `sck` domain?

File: hw/ip/spi_device/doc/_index.md [hunk 2]
```diff
@@ -410,6 +410,13 @@ Except BUSY bit, other bits are controlled by SW.
 BUSY bit is set by HW when it receives any commands that are uploaded to the FIFOs.
 SW may clear BUSY bit when it completes the received commands (e.g Erase/ Program).
 
+If BUSY is set, SPI_DEVICE IP blocks the passhthrough interface in Passthrough mode.
+The blocking of the interface occurs in SPI transaction idle state (CSb == 1).
+However, due to the instrinsic delay of the CDC, CSb transition is delayed by 2 SYS_CLK cycles.
+It may introduce a corner case when the host system starts sending a SPI transaction while SW may clear the BUSY signal.
+In that case, the blocking and unblocking of the passthrough may happen while SPI is active.
+The HW behavior in this scenario is not determined.
```
Comment (a-will) [inline hw/ip/spi_device/doc/_index.md:418]: This scenario can be avoided by observing the minimum CSb high time. (long enough to cover the max delay through the CDC)

And that's only true if changes are limited to the _rising edge_ of CSb (more specific trigger than CSb == 1). It looks like this is what you have implemented already in `spid_status`, for what it's worth. :)

File: hw/ip/spi_device/rtl/spi_device.sv [hunk 3]
```diff
@@ -1518,6 +1518,22 @@ module spi_device
...
+  // signal: sys_busy  -> output of u_status readstatus_d[0]
+  //              set by CSb deassertion pulse & BUSY(SCK)
+  //              clr by CSb = 1 & SW writing 0
+  //
+  // NOTE: there will be a gap between the actual assertion of CSb and the CSb
+  //   syncd event visible in the u_status BUSY logic (2FF @ SYS_CLK). So,
+  //   there's chance that the SW may clear the BUSY right at the CSb
+  //   assertion event. If that happens, passthrough block may set during SPI
+  //   transaction. The behavior of the SPI_DEVICE in this scenario is
+  //   undeterminstic.
```
Comment (a-will) [inline hw/ip/spi_device/rtl/spi_device.sv:1533]: Same note as above (in the doc) about requiring a minimum CSb high time to avoid that scenario.

**Review by a-will — APPROVED** [review]
Thanks for adding that! Now switching passthrough on/off will be much safer.

File: hw/ip/spi_device/data/spi_device.hjson (hunk 1 above)
Comment (eunchan) [reply hw/ip/spi_device/data/spi_device.hjson:521]: The update of the register in SYS_CLK domain is guarded by CSb (2FF synced version of). It updates only when CSb is high (set/clear).

So, in order to support SW to clear at anytime regardless of CSb state, HW should maintain two sets of BUSY. One is for the value to be updated, the other is to the indicator of the current BUSY @ SYS_CLK.

Let's think the case when SW clears the BUSY when SPI is active (CSb = 0). If we remove the limitation of CSb high condition, SW may clear the BUSY signal while SPI transaction is in active. Then, the passthrough will be un-blocked right away.

The BUSY @ SPI_CLK will be updated at the next SPI transaction though, as the design only updates BUSY @ SPI_CLK when CSb is asserted (0 -> 1). So, from the host system point of view, it won't raises the race condition.

To support safer operation, the logic gets complicated. I just wanted simpler logic inside, and read back of BUSY would not harm the SW much in my opinion.

File: hw/ip/spi_device/doc/_index.md (hunk 2 above)
Comment (eunchan) [reply hw/ip/spi_device/doc/_index.md:418]: yes

File: hw/ip/spi_device/rtl/spi_device.sv (hunk 3 above)
Comment (eunchan) [reply hw/ip/spi_device/rtl/spi_device.sv:1533]: QQ here. Could you please clarify the minimum CSb high time? I assume it is the host system responsibility, and SW may report the capability of the device via SFDP right? So, I assume HW does not have to check the minimum CSb high time.

Comment (a-will) [reply hw/ip/spi_device/rtl/spi_device.sv:1533]: This is something we'd specify in a data sheet, and the host must meet that spec. It is not discoverable (at least there's no mechanism for discovering it that I'm aware of).

The SFDP largely doesn't contain information about how to drive the SPI bus--It concerns itself with how to do the flash access application.

File: hw/ip/spi_device/data/spi_device.hjson (hunk 1 above)
Comment (a-will) [reply hw/ip/spi_device/data/spi_device.hjson:521]: Hm... I'm confused. I still don't understand why software cares about when the "clear event" hits the Status Register. The host's spi_host is the only one directly observing the BUSY bit in the Status Register, right? :thinking: 

There are three entities here:
1. Software
   - Software is only interested in issuing a "clear the BUSY bit" instruction to the spi_device, and this can be delayed. 
   - The CSR that represents the BUSY bit's state can update immediately as an option, if you want it to be used to indicate whether spi_host is blocked. However, this probably isn't a good use of that bit, so from software's perspective, this bit might as well be write-only.
   - Software generally should not hold itself up by waiting for the host to send another transaction (OK if that is actually the only job it has). That could come at any time, even hours later.
2. OT spi_host / passthrough
   - The passthrough mux should not change its selection while the spi_device has an active transaction.
   - But OT's spi_host _also_ should not be used while a "clear the BUSY bit" is pending. This intention can be tracked by a driver in RAM; software doesn't need to track the change via the CSR.
3. Host
   - Because we need to make changes to bits outside an active transaction (and we're crossing clock domains), we choose to change the Status Register at CSB posedge and specify a minimum CSB inactive time.

In other words, is there any reason software needs to read the BUSY bit from the CSR? It seems like software has done its job with the write.

Comment (eunchan) [reply hw/ip/spi_device/data/spi_device.hjson:521]: > Software is only interested in issuing a "clear the BUSY bit" instruction to the spi_device, and this can be delayed.
>
This is not true in the current design. It will be more complicated to support this behavior. And I didn't consider that at the beginning. If I wanted to do that. As I mentioned in the previous comment, the HW should maintain two copies of the BUSY (@ SYS_CLK) one for blocking the passthrough, another for the pending event of clearing the BUSY (@ SPI_CLK).

Comment (eunchan) [reply hw/ip/spi_device/data/spi_device.hjson:521]: Wait... Are you saying that the SW may wait until SPI transaction is completed? Then, yes SW does not have to read back to check if BUSY is cleared.

But I want to emphasize that the HW does not ensure the SW writes to the BUSY actually written.

Comment (a-will) [reply hw/ip/spi_device/data/spi_device.hjson:521]: > But I want to emphasize that the HW does not ensure the SW writes to the BUSY actually written.

That's what is probably where my confusion comes. You might want to word the doc to say that, since it's not clear what software's _must_ do (I was trying to ask "Why does software have to spend time confirming that the change happened?" heh).

Something like... "The clearing of the busy bit may not occur immediately. The completion happens at the end of a SPI transaction, and software may read the bit to confirm when the change occurs."

Then we're all good. :)

Comment (a-will) [reply hw/ip/spi_device/data/spi_device.hjson:521]: What was slightly confusing was what this statement meant:
```
SW should read back the register to confirm the value is cleared
```

With that construction, it's not totally clear that there is a conditional. It could be that SW should always read the register and wait for the clear to complete. Or it could be that SW should read the register _if it wants_ to observe the clear to complete--And I can't come up with a good reason why we'd really care to do that. It's fun ambiguity! ;)

Comment (a-will) [reply hw/ip/spi_device/data/spi_device.hjson:521]: > Uh-oh... I don't want to say this, but both are wrong. HW ignores any updates to the BUSY signal (from SW) when SPI transaction is active (CSb = 0). So, if SW read-back returns 1, SW shall write again when SPI is in idle.

Oh yeah, that doesn't sound ideal. Wouldn't you agree that software should not need to spin on a SPI transaction, retrying a clear until it's accepted? I'd recommend saving that SW intends to clear, then execute at the first opportunity.

**Comment by tjaychen** [comment]
this looks okay to me given all the discussion that was had, and the understanding that we will follow up with a different PR to address the corner cases pointed out. 

I think the general consensus now is to move towards using CSb as a clock signal to drive the eventual passthrough enable indication?

@a-will @eunchan does that sound right to you guys?

**Review by tjaychen — APPROVED** [review]

**Comment by eunchan** [comment]
Thanks for the review Tim. I will update the following PR accordingly.
